File: Calculator/ios-native/CalculatorHub/Calculators/GeometryCalculators.swift
import Foundation

// Geometry Calculators

// Circle Properties
struct CircleResult { let circumference: Double; let area: Double; let diameter: Double }
class CircleCalculator {
    func fromRadius(_ r: Double) -> CircleResult { CircleResult(circumference: 2 * .pi * r, area: .pi * r * r, diameter: 2 * r) }
    func fromDiameter(_ d: Double) -> CircleResult { CircleResult(circumference: .pi * d, area: .pi * pow(d/2, 2), diameter: d) }
}

// Slope Calculator
struct SlopeResult { let slope: Double; let yIntercept: Double; let equation: String; let distance: Double }
class SlopeCalculator {
    func calculate(x1: Double, y1: Double, x2: Double, y2: Double) -> SlopeResult {
        let slope = (y2 - y1) / (x2 - x1); let yInt = y1 - slope * x1
        let dist = sqrt(pow(x2 - x1, 2) + pow(y2 - y1, 2))
        return SlopeResult(slope: slope, yIntercept: yInt, equation: "y = \(slope)x + \(yInt)", distance: dist)
    }
}

// Triangle Calculator
struct TriangleResult { let area: Double; let perimeter: Double; let hypotenuse: Double? }
class TriangleCalculator {
    func fromBaseHeight(base: Double, height: Double) -> TriangleResult { TriangleResult(area: 0.5 * base * height, perimeter: 0, hypotenuse: nil) }
    func fromSides(a: Double, b: Double, c: Double) -> TriangleResult {
        let s = (a + b + c) / 2; let area = sqrt(s * (s-a) * (s-b) * (s-c))
        return TriangleResult(area: area, perimeter: a + b + c, hypotenuse: nil)
    }
    func pythagorean(a: Double, b: Double) -> TriangleResult { TriangleResult(area: 0.5 * a * b, perimeter: a + b + sqrt(a*a + b*b), hypotenuse: sqrt(a*a + b*b)) }
}

// Volume Calculator
struct VolumeResult { let volume: Double; let surfaceArea: Double }
class VolumeCalculator {
    func cube(side: Double) -> VolumeResult { VolumeResult(volume: pow(side, 3), surfaceArea: 6 * pow(side, 2)) }
    func sphere(radius: Double) -> VolumeResult { VolumeResult(volume: (4.0/3.0) * .pi * pow(radius, 3), surfaceArea: 4 * .pi * pow(radius, 2)) }
    func cylinder(radius: Double, height: Double) -> VolumeResult { VolumeResult(volume: .pi * pow(radius, 2) * height, surfaceArea: 2 * .pi * radius * (radius + height)) }
    func cone(radius: Double, height: Double) -> VolumeResult { VolumeResult(volume: (1.0/3.0) * .pi * pow(radius, 2) * height, surfaceArea: .pi * radius * (radius + sqrt(pow(height, 2) + pow(radius, 2)))) }
    func box(l: Double, w: Double, h: Double) -> VolumeResult { VolumeResult(volume: l * w * h, surfaceArea: 2 * (l*w + w*h + h*l)) }
}

// Surface Area Calculator
class SurfaceAreaCalculator {
    func cubeSA(side: Double) -> Double { 6 * pow(side, 2) }
    func sphereSA(radius: Double) -> Double { 4 * .pi * pow(radius, 2) }
    func cylinderSA(radius: Double, height: Double) -> Double { 2 * .pi * radius * (radius + height) }
}

// Aspect Ratio Calculator
struct AspectRatioResult { let ratio: String; let width: Double; let height: Double }
class AspectRatioCalculator {
    private func gcd(_ a: Int, _ b: Int) -> Int { b == 0 ? a : gcd(b, a % b) }
    func calculate(width: Double, height: Double) -> AspectRatioResult {
        let g = gcd(Int(width), Int(height))
        return AspectRatioResult(ratio: "\(Int(width)/g):\(Int(height)/g)", width: width, height: height)
    }
}

// Fraction Calculator
struct FractionResult { let numerator: Int; let denominator: Int; let decimal: Double; let simplified: String }
class FractionCalculator {
    private func gcd(_ a: Int, _ b: Int) -> Int { b == 0 ? abs(a) : gcd(b, abs(a) % abs(b)) }
    func simplify(num: Int, den: Int) -> FractionResult {
        let g = gcd(num, den); let n = num / g; let d = den / g
        return FractionResult(numerator: n, denominator: d, decimal: Double(num) / Double(den), simplified: "\(n)/\(d)")
    }
    func add(n1: Int, d1: Int, n2: Int, d2: Int) -> FractionResult { simplify(num: n1 * d2 + n2 * d1, den: d1 * d2) }
    func multiply(n1: Int, d1: Int, n2: Int, d2: Int) -> FractionResult { simplify(num: n1 * n2, den: d1 * d2) }
}

// Permutation & Combination
class PermCombCalculator {
    private func factorial(_ n: Int) -> Double { (2...max(n, 1)).reduce(1.0) { $0 * Double($1) } }
    func permutation(_ n: Int, _ r: Int) -> Double { factorial(n) / factorial(n - r) }
    func combination(_ n: Int, _ r: Int) -> Double { factorial(n) / (factorial(r) * factorial(n - r)) }
}

File: Calculator/android-native/app/src/main/java/com/qaantum/calculatorhub/calculators/GeometryCalculators.kt
package com.qaantum.calculatorhub.calculators

import kotlin.math.*

// Geometry Calculators

// Circle Properties
data class CircleResult(val circumference: Double, val area: Double, val diameter: Double)
class CircleCalculator {
    fun fromRadius(r: Double) = CircleResult(2 * PI * r, PI * r * r, 2 * r)
    fun fromDiameter(d: Double) = CircleResult(PI * d, PI * (d/2) * (d/2), d)
    fun fromCircumference(c: Double): CircleResult { val r = c / (2 * PI); return CircleResult(c, PI * r * r, 2 * r) }
}

// Slope Calculator
data class SlopeResult(val slope: Double, val yIntercept: Double, val equation: String, val distance: Double)
class SlopeCalculator {
    fun calculate(x1: Double, y1: Double, x2: Double, y2: Double): SlopeResult {
        val slope = (y2 - y1) / (x2 - x1)
        val yInt = y1 - slope * x1
        val dist = sqrt((x2 - x1).pow(2) + (y2 - y1).pow(2))
        return SlopeResult(slope, yInt, "y = ${slope}x + $yInt", dist)
    }
}

// Triangle Calculator
data class TriangleResult(val area: Double, val perimeter: Double, val hypotenuse: Double?)
class TriangleCalculator {
    fun fromBasHeight(base: Double, height: Double) = TriangleResult(0.5 * base * height, 0.0, null)
    fun fromSides(a: Double, b: Double, c: Double): TriangleResult {
        val s = (a + b + c) / 2
        val area = sqrt(s * (s-a) * (s-b) * (s-c)) // Heron's formula
        return TriangleResult(area, a + b + c, null)
    }
    fun pythagorean(a: Double, b: Double) = TriangleResult(0.5 * a * b, a + b + sqrt(a*a + b*b), sqrt(a*a + b*b))
}

// Volume Calculator
data class VolumeResult(val volume: Double, val surfaceArea: Double)
class VolumeCalculator {
    fun cube(side: Double) = VolumeResult(side.pow(3), 6 * side.pow(2))
    fun sphere(radius: Double) = VolumeResult((4.0/3.0) * PI * radius.pow(3), 4 * PI * radius.pow(2))
    fun cylinder(radius: Double, height: Double) = VolumeResult(PI * radius.pow(2) * height, 2 * PI * radius * (radius + height))
    fun cone(radius: Double, height: Double) = VolumeResult((1.0/3.0) * PI * radius.pow(2) * height, PI * radius * (radius + sqrt(height.pow(2) + radius.pow(2))))
    fun box(l: Double, w: Double, h: Double) = VolumeResult(l * w * h, 2 * (l*w + w*h + h*l))
}

// Surface Area Calculator
class SurfaceAreaCalculator {
    fun cubeSA(side: Double) = 6 * side.pow(2)
    fun sphereSA(radius: Double) = 4 * PI * radius.pow(2)
    fun cylinderSA(radius: Double, height: Double) = 2 * PI * radius * (radius + height)
    fun coneSA(radius: Double, slantHeight: Double) = PI * radius * (radius + slantHeight)
}

// Aspect Ratio Calculator
data class AspectRatioResult(val ratio: String, val width: Double, val height: Double)
class AspectRatioCalculator {
    private fun gcd(a: Int, b: Int): Int = if (b == 0) a else gcd(b, a % b)
    fun calculate(width: Double, height: Double): AspectRatioResult {
        val g = gcd(width.toInt(), height.toInt())
        return AspectRatioResult("${(width/g).toInt()}:${(height/g).toInt()}", width, height)
    }
    fun scaleToWidth(originalW: Double, originalH: Double, newW: Double): AspectRatioResult {
        val scale = newW / originalW
        return AspectRatioResult("", newW, originalH * scale)
    }
}

// Fraction Calculator
data class FractionResult(val numerator: Int, val denominator: Int, val decimal: Double, val simplified: String)
class FractionCalculator {
    private fun gcd(a: Int, b: Int): Int = if (b == 0) abs(a) else gcd(b, abs(a) % abs(b))
    
    fun simplify(num: Int, den: Int): FractionResult {
        val g = gcd(num, den)
        val n = num / g; val d = den / g
        return FractionResult(n, d, num.toDouble() / den, "$n/$d")
    }
    fun add(n1: Int, d1: Int, n2: Int, d2: Int): FractionResult {
        val num = n1 * d2 + n2 * d1; val den = d1 * d2
        return simplify(num, den)
    }
    fun multiply(n1: Int, d1: Int, n2: Int, d2: Int) = simplify(n1 * n2, d1 * d2)
    fun divide(n1: Int, d1: Int, n2: Int, d2: Int) = simplify(n1 * d2, d1 * n2)
}

// Permutation & Combination
class PermCombCalculator {
    private fun factorial(n: Int): Double { var r = 1.0; for (i in 2..n) r *= i; return r }
    fun permutation(n: Int, r: Int): Double = factorial(n) / factorial(n - r)
    fun combination(n: Int, r: Int): Double = factorial(n) / (factorial(r) * factorial(n - r))
}
